### DISCUSSION TITLE
How to use `column.op()` for `IN`?

### DISCUSSION THREAD

**Discussion by nexushoratio** [sqlalchemy/sqlalchemy#11934] [Usage Questions] [answered]
Still on 1.4.46.

I am building select statements up on the fly, and at the point where I can do something like this:
```python
stmt = stmt.where(field_map[field].bool_op(op_str)(value))
```
Where `field_map` is a preprocessed collection from `stmt.exported_columns`.

This works fine for common operators like `>`, `>=`, and so on.

But I have not figured out if there is an equivalent for `IN` and `NOT IN`.  Or rather, what the value for `op_str` should  be.

I can achieve similar with:
```python
my_op = getattr(field_map[field], 'in_')
stmt = stmt.where(my_op(values))
```

If there is a more official way of achieving this, however, I would prefer to do that.

Any guidance?

**Comment by zzzeek** [comment, marked as answer]
ok you dont want to do either of those ways, in Python operators are represented by functions:

```py
>>> import operator
>>> operator.add(3, 5)
8
```

SQLAlchemy has these all listed out in https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/operators.py .   these are the actual operators, and you want to use those.   link them up to strings:

```py

from sqlalchemy.sql import operators

lookup = {
    "<": operators.lt,
    "<=": operators.le, 
    "+": operators.add,
    # ...

    "IN": operators.in_op,
    "NOT IN": operators.not_in_op,

}
```

Using `.op()` is going to lead to poor results as you will not get any proper operator precedence behaviors and as you note with IN there are lots of operators that have special behaviors which you wont get with `op()`.

**Reply by nexushoratio** [reply]
Thanks.

I ended up actually using a map to Python's `operator` module instead.

It may end up with an extra layer of redirection as `operator.X` ends up calling `sqlalchemy.sql.operators.X`, but for some reason, `mypy` seems unhappy with that module:
```
error: Module has no attribute "gt"  [attr-defined] 
```

I'll just chalk that up to my instanced on preferring to use code as delivered by Debian rather than using pip, and thus being a bit out of date.

**Reply by zzzeek** [reply]
likely sqlalchemy2-stubs doesn't have these functions.    Overall, if you're writing new code I would really strongly recommend you get onto 2.0 first.   Getting mypy to work with 1.4 is a very uphill stream to travel
